File: sdks/typescript/src/v1/dot-segment-guard.ts
// Wraps `ServerConfiguration.makeRequestContext`, the one seam that still
// holds the raw path string before `new URL(...)` resolves a ".." segment
// away; a `Middleware.pre()` hook runs too late for that.
import type { BaseServerConfiguration } from "./generated/servers.js";
import type { HttpMethod, RequestContext } from "./generated/http/http.js";
import { E2AValidationError } from "./errors.js";

function dotSegment(endpoint: string): string | undefined {
  return endpoint.split("/").find((segment) => /^\.+$/.test(segment));
}

/** Wraps a `BaseServerConfiguration` to reject any outgoing request whose
 *  path contains a dot-only segment ("." or "..") before it can be resolved
 *  away by URL parsing. */
export function guardDotSegments(inner: BaseServerConfiguration): BaseServerConfiguration {
  return {
    makeRequestContext(endpoint: string, httpMethod: HttpMethod): RequestContext {
      const bad = dotSegment(endpoint);
      if (bad !== undefined) {
        throw new E2AValidationError({
          code: "invalid_request_path",
          message: `refusing to send a request whose path contains a dot-segment ("${bad}"): ${endpoint}`,
          status: 0,
          retryable: false,
        });
      }
      return inner.makeRequestContext(endpoint, httpMethod);
    },
  };
}

File: sdks/python/src/e2a/v1/_dot_segment_guard.py
"""Reject a dot-segment path parameter before httpx can resolve it as "go up
one level" and retarget the request onto a parent resource. Called from
``_TypedApiClient.param_serialize`` in ``client.py``, on the raw
``path_params`` dict, before the path string is built."""

from __future__ import annotations

import re
from typing import Mapping, Optional

from .errors import E2AValidationError

_ALL_DOTS = re.compile(r"^\.+$")


def reject_dot_segment_path_params(path_params: Optional[Mapping[str, object]]) -> None:
    if not path_params:
        return
    for name, value in path_params.items():
        if isinstance(value, str) and _ALL_DOTS.match(value):
            raise E2AValidationError(
                code="invalid_request_path",
                message=(
                    f'refusing to send a request whose {name!r} path parameter '
                    f'is a dot-segment ("{value}")'
                ),
                status=0,
                retryable=False,
            )
